src/lib/gossip_net/gossip_net.ml

```diff
@@ -239,25 +240,40 @@ module Make (Message : Message_intf) :
                     "%s" (Exn.to_string_mach exn) ))
             (Tcp.Where_to_listen.of_port config.me.Peer.communication_port)
             (fun client reader writer ->
-              let%map _ =
-                Rpc.Connection.server_with_close reader writer ~implementations
-                  ~connection_state:(fun conn ->
-                    (* connection state is the client's IP and ephemeral port
+              let conn_list =
+                Option.value_map ~default:[]
+                  (Hashtbl.find t.connections (Socket.Address.Inet.addr client))
+                  ~f:Fn.id
+              in
+              let connection_limit = List.length implementation_list + 1 in
+              if List.length conn_list = connection_limit then (
+                Logger.error t.logger ~module_:__MODULE__ ~location:__LOC__
+                  "Cannot open another connection. Number of existing \
+                   connections from %s equals the limit %d."
+                  (Socket.Address.Inet.to_string client)
+                  connection_limit ;
+                Deferred.unit )
+              else
+                let%map _ =
+                  Rpc.Connection.server_with_close reader writer
+                    ~implementations
+                    ~connection_state:(fun conn ->
+                      (* connection state is the client's IP and ephemeral port
                         when connecting to the server over TCP; the ephemeral
                         port is distinct from the client's discovery and
                         communication ports *)
-                    Hashtbl.set t.connections
-                      ~key:(Socket.Address.Inet.addr client)
-                      ~data:conn ;
-                    Socket.Address.Inet.to_host_and_port client )
-                  ~on_handshake_error:
-                    (`Call
-                      (fun exn ->
-                        Logger.error t.logger ~module_:__MODULE__
-                          ~location:__LOC__ "%s" (Exn.to_string_mach exn) ;
-                        Deferred.unit ))
-              in
-              Hashtbl.remove t.connections @@ Socket.Address.Inet.addr client
+                      Hashtbl.set t.connections
+                        ~key:(Socket.Address.Inet.addr client)
+                        ~data:(conn :: conn_list) ;
+                      Socket.Address.Inet.to_host_and_port client )
+                    ~on_handshake_error:
+                      (`Call
+                        (fun exn ->
+                          Logger.error t.logger ~module_:__MODULE__
+                            ~location:__LOC__ "%s" (Exn.to_string_mach exn) ;
+                          Deferred.unit ))
+                in
+                Hashtbl.remove t.connections @@ Socket.Address.Inet.addr client
```

Comment:
This removes all the incoming connections from the hashtable. You need to only remove the current one. Probably use a set rather than a list here.